.changeset/empty-fans-reflect.md

```diff
@@ -0,0 +1,70 @@
+---
+'astro': minor
+---
+
+Adds a new `experimental.tsconfig` configuration option
+
+If you enable this option, Astro will generate a `tsconfig.json` file under `./.astro/`:
+
+```diff
+import { defineConfig } from "astro/config"
+
+export default defineConfig({
++    experimental: {
++        tsconfig: {}
++    }
+})
+```
+
+It will create references under the hood, making `env.d.ts` optional if you haven't made any change to it:
+
+```diff
+-/// <reference types="astro/client" />
+-/// <reference path="../.astro/types.d.ts" />
+```
+
+Enabling this option will require a few updates to your root `tsconfig.json`:
+
+1. `extends` must point to `./.astro/tsconfig.json` otherwise an error will be thrown with a diff to help you.
+2. If set, move `include` and `exclude` to `experimental.tsconfig.include` and `experimental.tsconfig.exclude` respectively in your Astro config file.
+
+For example, in `tsconfig.json` you'll need to make the following changes:
+
+```diff
+{
++    "extends": "./.astro/tsconfig.json"
+-    "include": ["foo"],
+-    "exclude": ["bar"]
+}
+```
+
+Then you'll need to reflect them in `astro.config.mjs`:
+
+```diff
+import { defineConfig } from "astro/config"
+
+export default defineConfig({
+    experimental: {
+        tsconfig: {
++            include: ["foo"],
++            exclude: ["bar"]
+        }
+    }
+})
+```
+
+By default, `outDir` will now be excluded to avoid a long standing issue when using `astro check` on an already built project. If you want to opt-out of this behavior, you can set `experimental.typescript.excludeOutDir` to `false`:
```

Comment:
Is there a discussion about needing `excludeOutDir`? I think we should be strict here and not introduce that option. It's unlikely that a user wants to not exclude the out dir I think

Reply:
I don't remember exactly where/when it occurred but @Princesseuh told me some people actually want to check their build folder for some reason, although it's rare

Reply:
Yea, some people once wanted it, I think it could be excluded and those people wouldn't necessarily lose sleep over it, though. We can reconsider if someone else ask